src/core/api.ts

```diff
@@ -6,15 +6,22 @@ import { GET_CHROM_SIZES } from './utils/assembly';
 import { CompleteThemeDeep } from './utils/theme';
 import { traverseViewsInViewConfig } from './utils/view-config';
 
-export type CommonEventData = {
-    genomicPosition: string;
-    data: Datum[];
-};
-
-export type RawDataEventData = {
+export interface CommonEventData {
+    /** Source visualization ID, i.e., `track.id` */
     id: string;
+    /** Values in a JSON array that represent data after data transformation */
     data: Datum[];
-};
+}
+
+export interface PointMouseEventData extends CommonEventData {
+    /* A genomic coordinate, e.g., `chr1:100,000`. */
+    genomicPosition: string;
+}
+
+export interface RangeMouseEventData extends CommonEventData {
+    /* Start and end genomic coordinates, e.g., `chr1:100,000`. NULL if a range is deselected. */
+    genomicRange: [string, string] | null;
```

Comment:
Similar thought here, `string`s are probably useful for a human reading but not for a consumer of the API. Consumers would have to know how strings are formatted in-order to parse into something more useable.

However, I also don't think a pair of locations is enough information either for range. E.g.,

```typescript
type GenomicLocation = [chrom: string, position: number]; 
type GenomicRange = [GenomicLocation, GenomicLocation];

interface RangeMouseEventData extends CommonEventData {
  genomicRange: GenomicRange | null;
}
```

single a `GenomicRange` could span multiple chromosomes.

For example, returning `["chr1", 100000]` and `["chr6", 10]` isn't very helpful since an end-users will need to look up individual chromosome lengths and determine start/ends of each chromosome within this interval _in Gosling_.

An genomic assembly only defines sequence composition _within_ chromosome, and Gosling decides that chromosomes be ordered in a particular way end-to-end. Let's say we have a different chromosome ordering in Gosling (e.g. (`'chr1', 'chr6', 'chr2', chr3', ...`)), then the  genomic range `["chr1", 100000]` to `["chr6", 10]` means something entirely different. Of course chromosomes with numbered names have a natural order, but that's totally arbitrary.

I think we could improve this API by returning data for genomic range within each _chromosome_ selected:

```typescript
type GenomicInterval = [chrom: name, left: number, right: number];

interface RangeMouseEventData {
    id: string;
    intervals: { interval: GenomicInterval, data: Datum[] }[],
}
```

so an interval spanning chr1, chr2, chr3 would return something like:

```typescript
// [
//   { interval: ['chr1', 100000, 248956422], data: [/* ... */] },
//   { interval: ['chr2', 0, 242193529], data: [/* ... */] },
//   { interval: ['chr3', 0, 20000], data: [/* ... */] },
// ]
```

Reply:
I like the idea! Given that we are planning to support custom chromosome sizes (#582), this approach is making more sense to me. Consumers can still easily make a flat array of `data` if they want.

Regarding the naming and structure, what do you think of the following, which is more consistent with `DomainChrInterval` in our grammar?

```js
// [
//   { chromosome: 'chr1', interval: [100000, 248956422], data: [/* ... */] },
//   { chromosome: 'chr2', interval: [0, 242193529], data: [/* ... */] },
//   { chromosome: 'chr3', interval: [0, 20000], data: [/* ... */] },
// ]
```

```ts
// gosling.schema.ts
export interface DomainChrInterval {
    chromosome: Chromosome;
    interval: [number, number];
}
```

Reply:
> Regarding the naming and structure, what do you think of the following, which is more consistent with `DomainChrInterval` in our grammar?

I think this is a great idea!

Reply:
I find that we need to consider cases where marks do not correspond to a single chromosome. For example, a user can select a link that connects between `chr1` and `chr7`.

We can perhaps include the same value in both `chromosome: 'chr1', data: [...]` and `chromosome: 'chr7', data: [...]`, but API consumers will need to remove the duplication by themselves if they want a flat array.

Unsure at the moment what method would be better between (1) a flat array vs. (2) per-chromosome arrays. The flat array is simpler in terms of the structure, and consumers can still get per-chromosome arrays by looking up raw data fields (e.g., a `chr` column).

Reply:
That's a good point, and something I had not thought of. It would be more simple to return the flat array as you suggested, since these correspond to the "selected" entries in the data. I wonder if we could do something similar to what you had before then,

```typescript
type GenomicRegion = [chr: string, start: number, end: number];

export interface RangeMouseEventData extends CommonEventData {
  interval: [GenomicRegion, GenomicRegion],
}
```

I guess users have enough information to understand which chromosomes are within this `interval` from the assembly (chromsizes).